File: libs/ui_wellseis/src/qt/seismic_context_toolbar.cpp
#include <pwb/ui_wellseis/qt/seismic_context_toolbar.hpp>

#include <QAction>
#include <QActionGroup>
#include <QComboBox>
#include <QGridLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QMenu>
#include <QPushButton>
#include <QToolButton>
#include <QWidgetAction>

#include <pwb/ui_wellseis/output_labels.hpp>
#include <pwb/ui_wellseis/seismic_attributes.hpp>

namespace pwb::ui_wellseis::qt {

namespace {

QString qs(const std::string& text) {
    return QString::fromUtf8(text.data(), static_cast<int>(text.size()));
}

QLabel* field_label(const QString& text, QWidget* parent) {
    auto* label = new QLabel(text, parent);
    label->setObjectName(QStringLiteral("WorkFieldLabel"));
    return label;
}

QLabel* field_value_label(const QString& text, QWidget* parent) {
    auto* label = new QLabel(text, parent);
    label->setObjectName(QStringLiteral("WorkFieldValue"));
    return label;
}

}  // namespace

SeismicContextToolbar::SeismicContextToolbar(QWidget* parent)
    : QFrame(parent) {
    setObjectName(QStringLiteral("SeismicContextToolbar"));
    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(12, 4, 12, 4);
    layout->setSpacing(8);

    // 1. 工区地震体 (source volume selector)
    layout->addWidget(field_label(QStringLiteral("工区地震体"), this));
    source_combo_ = new QComboBox(this);
    source_combo_->setObjectName(QStringLiteral("SeismicPredictionSourceCombo"));
    source_combo_->setPlaceholderText(
        QStringLiteral("选择数据管理中的 SEG-Y 地震体"));
    source_combo_->setToolTip(QStringLiteral(
        "选择工区数据中已归档的 .sgy / .segy 地震体，加载后可直接运行预测"));
    source_combo_->setMinimumWidth(180);
    connect(source_combo_, &QComboBox::currentIndexChanged, this, [this](int) {
        emit source_changed(
            source_combo_->currentData().toString());
    });
    layout->addWidget(source_combo_, 1);

    // 2. 地震属性下拉
    layout->addWidget(field_label(QStringLiteral("属性"), this));
    attribute_combo_ = new QComboBox(this);
    attribute_combo_->setObjectName(
        QStringLiteral("SeismicAttributeDropdownCombo"));
    attribute_combo_->setToolTip(QStringLiteral("快速切换当前显示的地震属性"));
    for (const std::string& attribute : all_seismic_attributes()) {
        attribute_combo_->addItem(qs(attribute));
    }
    connect(attribute_combo_, &QComboBox::currentTextChanged, this,
            [this](const QString& text) {
                if (!suppress_signals_ && !text.isEmpty()) {
                    attribute_value_->setText(text);
                    emit attribute_changed(text);
                }
            });
    layout->addWidget(attribute_combo_);

    // 3. 设置与详情 popover
    settings_btn_ = new QToolButton(this);
    settings_btn_->setObjectName(QStringLiteral("SecondaryButton"));
    settings_btn_->setText(QStringLiteral("设置与详情 ▾"));
    settings_btn_->setToolTip(
        QStringLiteral("查看预测任务详情、目标层位、显示模式与状态"));
    settings_btn_->setPopupMode(QToolButton::InstantPopup);
    build_settings_menu();
    layout->addWidget(settings_btn_);

    // 4. 状态
    layout->addWidget(field_label(QStringLiteral("状态:"), this));
    status_value_ = field_value_label(QStringLiteral("—"), this);
    layout->addWidget(status_value_);

    layout->addStretch(1);

    // 6. 操作按钮
    demo_btn_ = new QPushButton(QStringLiteral("运行演示预测"), this);
    demo_btn_->setObjectName(QStringLiteral("SecondaryButton"));
    demo_btn_->setToolTip(QStringLiteral(
        "显式演示模式：运行 DemoModelProvider（合成数据，非科学预测）"));
    connect(demo_btn_, &QPushButton::clicked, this,
            &SeismicContextToolbar::demo_requested);
    layout->addWidget(demo_btn_);

    run_btn_ = new QPushButton(QStringLiteral("运行预测"), this);
    run_btn_->setObjectName(QStringLiteral("PrimaryButton"));
    run_btn_->setToolTip(
        QStringLiteral("通过 ModelRegistry 解析生产模型后运行科学预测；"
                       "未配置生产模型时不会自动运行 mock"));
    connect(run_btn_, &QPushButton::clicked, this,
            &SeismicContextToolbar::run_requested);
    layout->addWidget(run_btn_);
}

void SeismicContextToolbar::build_settings_menu() {
    settings_menu_ = new QMenu(this);
    settings_menu_->setObjectName(QStringLiteral("SeismicSettingsMenu"));

    auto* details = new QFrame(settings_menu_);
    details->setObjectName(QStringLiteral("SeismicSettingsDetailsCard"));
    auto* grid = new QGridLayout(details);
    grid->setContentsMargins(8, 4, 8, 4);
    grid->setSpacing(4);

    grid->addWidget(field_label(QStringLiteral("任务:"), details), 0, 0);
    task_value_ =
        field_value_label(QStringLiteral("未选择预测任务"), details);
    grid->addWidget(task_value_, 0, 1);

    grid->addWidget(field_label(QStringLiteral("层位:"), details), 1, 0);
    horizon_value_ = field_value_label(QStringLiteral("—"), details);
    grid->addWidget(horizon_value_, 1, 1);

    grid->addWidget(field_label(QStringLiteral("当前属性:"), details), 2, 0);
    attribute_value_ = field_value_label(QStringLiteral("振幅"), details);
    grid->addWidget(attribute_value_, 2, 1);

    grid->addWidget(field_label(QStringLiteral("显示模式:"), details), 3, 0);
    mode_value_ = field_value_label(QStringLiteral("vd"), details);
    grid->addWidget(mode_value_, 3, 1);

    grid->addWidget(field_label(QStringLiteral("体数据维度:"), details), 4, 0);
    shape_value_ = field_value_label(QStringLiteral("—"), details);
    grid->addWidget(shape_value_, 4, 1);

    grid->addWidget(field_label(QStringLiteral("输出性质:"), details), 5, 0);
    mock_value_ = field_value_label(QStringLiteral("—"), details);
    grid->addWidget(mock_value_, 5, 1);

    auto* card_action = new QWidgetAction(settings_menu_);
    card_action->setDefaultWidget(details);
    settings_menu_->addAction(card_action);
    settings_menu_->addSeparator();

    // Display-mode submenu (checkable group).
    QMenu* mode_menu =
        settings_menu_->addMenu(QStringLiteral("切换显示模式"));
    mode_group_ = new QActionGroup(this);
    for (const std::string& mode : seismic_display_modes()) {
        QAction* action = new QAction(qs(mode), this);
        action->setCheckable(true);
        if (mode == kDisplayModeVd) {
            action->setChecked(true);
        }
        const QString mode_name = qs(mode);
        connect(action, &QAction::triggered, this, [this, mode_name] {
            mode_value_->setText(mode_name);
            emit display_mode_changed(mode_name);
        });
        mode_group_->addAction(action);
        mode_menu->addAction(action);
    }

    // Attribute submenu grouped by catalog.
    QMenu* attr_menu =
        settings_menu_->addMenu(QStringLiteral("切换地震属性"));
    for (const SeismicAttributeGroup& group : seismic_attribute_groups()) {
        QMenu* group_menu = attr_menu->addMenu(qs(group.group_label));
        for (const std::string& label : group.attributes) {
            QAction* action = new QAction(qs(label), this);
            const QString name = qs(label);
            connect(action, &QAction::triggered, this, [this, name] {
                set_selected_attribute(name);
                emit attribute_changed(name);
            });
            group_menu->addAction(action);
        }
    }

    settings_btn_->setMenu(settings_menu_);
}

void SeismicContextToolbar::set_context(
    const PredictionTaskSlice* task,
    const std::string& horizon,
    const std::string& attribute_label,
    const std::string& mode_label,
    std::optional<std::array<std::int64_t, 3>> volume_shape,
    const std::optional<std::string>& mock_nature) {
    task_value_->setText(
        task != nullptr && !task->name.empty()
            ? qs(task->name)
            : QStringLiteral("未选择预测任务"));
    horizon_value_->setText(horizon.empty() ? QStringLiteral("—")
                                            : qs(horizon));
    set_selected_attribute(qs(attribute_label_or_default(attribute_label)));
    const QString mode =
        mode_label.empty() ? QStringLiteral("vd") : qs(mode_label);
    mode_value_->setText(mode);
    for (QAction* action : mode_group_->actions()) {
        if (action->text() == mode) {
            action->setChecked(true);
        }
    }
    if (volume_shape.has_value()) {
        shape_value_->setText(qs(volume_shape_text(*volume_shape)));
    }
    if (mock_nature.has_value()) {
        mock_value_->setText(qs(*mock_nature));
    }
}

void SeismicContextToolbar::set_status(const QString& text) {
    status_value_->setText(text.isEmpty() ? QStringLiteral("—") : text);
}

void SeismicContextToolbar::set_inferring(bool inferring) {
    run_btn_->setEnabled(!inferring);
    demo_btn_->setEnabled(!inferring);
    if (inferring) {
        status_value_->setText(QStringLiteral("推断中…"));
    }
}

void SeismicContextToolbar::set_source_entries(
    const std::vector<SourceComboEntry>& entries) {
    const QString previous = source_combo_->currentData().toString();
    const QSignalBlocker blocker(source_combo_);
    source_combo_->clear();
    for (const SourceComboEntry& entry : entries) {
        source_combo_->addItem(qs(entry.label), qs(entry.resource_id));
    }
    const int index = previous.isEmpty()
                          ? -1
                          : source_combo_->findData(previous);
    // Python setCurrentIndex(selected_index) parity: an unresolved previous
    // selection lands on -1 (nothing shown), never silently on row 0.
    source_combo_->setCurrentIndex(index);
}

std::string SeismicContextToolbar::selected_source_id() const {
    return source_combo_->currentData().toString().toStdString();
}

void SeismicContextToolbar::set_selected_attribute(const QString& label) {
    const QString text = label.trimmed();
    if (text.isEmpty()) {
        return;
    }
    attribute_value_->setText(text);
    suppress_signals_ = true;
    const int index = attribute_combo_->findText(text);
    if (index >= 0) {
        attribute_combo_->setCurrentIndex(index);
    }
    suppress_signals_ = false;
}

void SeismicContextToolbar::set_display_mode(const QString& mode) {
    mode_value_->setText(mode.isEmpty() ? QStringLiteral("vd") : mode);
    for (QAction* action : mode_group_->actions()) {
        if (action->text() == mode_value_->text()) {
            action->setChecked(true);
        }
    }
}

}  // namespace pwb::ui_wellseis::qt

File: legacy/python_reference/product/paleo_workbench/ui/pages/seismic_context_toolbar.py
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtGui import QAction, QActionGroup
from PySide6.QtWidgets import (
    QComboBox,
    QFrame,
    QGridLayout,
    QHBoxLayout,
    QLabel,
    QMenu,
    QPushButton,
    QToolButton,
    QVBoxLayout,
    QWidget,
    QWidgetAction,
)

from paleo_workbench.ui import tokens
from paleo_workbench.viz.prediction_helpers import field_value

_ATTRIBUTE_GROUPS = (
    ("振幅属性", ("振幅", "包络", "RMS振幅")),
    ("频率属性", ("瞬时频率",)),
    ("连续性属性", ("甜点", "相对阻抗")),
    ("结构属性", ("瞬时相位", "Dip_IL", "Dip_XL", "方位角", "平均曲率", "高斯曲率", "最大曲率")),
    ("多属性融合", ("RGB融合",)),
)

ALL_SEISMIC_ATTRIBUTES = [
    attr for _group, attrs in _ATTRIBUTE_GROUPS for attr in attrs
]


class SeismicContextToolbar(QFrame):
    """Single-row compact context toolbar with dropdown popovers for seismic prediction."""

    run_requested = Signal()
    demo_requested = Signal()
    attribute_changed = Signal(str)
    display_mode_changed = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("SeismicContextToolbar")
        self._inferring = False
        self._suppress_signals = False

        layout = QHBoxLayout(self)
        layout.setContentsMargins(
            tokens.PANEL_PADDING,
            tokens.SPACE_2,
            tokens.PANEL_PADDING,
            tokens.SPACE_2,
        )
        layout.setSpacing(tokens.SPACE_3)

        # 1. 工区地震体 (Source Volume Selector)
        source_label = QLabel("工区地震体")
        source_label.setObjectName("WorkFieldLabel")
        layout.addWidget(source_label)

        self.seismic_source_combo = QComboBox()
        self.seismic_source_combo.setObjectName("SeismicPredictionSourceCombo")
        self.seismic_source_combo.setPlaceholderText("选择数据管理中的 SEG-Y 地震体")
        self.seismic_source_combo.setToolTip(
            "选择工区数据中已归档的 .sgy / .segy 地震体，加载后可直接运行预测"
        )
        self.seismic_source_combo.setMinimumWidth(180)
        layout.addWidget(self.seismic_source_combo, 1)

        # 2. 地震属性下拉 (Seismic Attribute Selector)
        attr_label = QLabel("属性")
        attr_label.setObjectName("WorkFieldLabel")
        layout.addWidget(attr_label)

        self.attribute_combo = QComboBox()
        self.attribute_combo.setObjectName("SeismicAttributeDropdownCombo")
        self.attribute_combo.setToolTip("快速切换当前显示的地震属性")
        self.attribute_combo.addItems(ALL_SEISMIC_ATTRIBUTES)
        self.attribute_combo.currentTextChanged.connect(self._on_attribute_combo_changed)
        layout.addWidget(self.attribute_combo)

        # 3. 设置与属性弹出下拉按钮 (Settings & Context Dropdown Popover)
        self.settings_btn = QToolButton()
        self.settings_btn.setObjectName("SecondaryButton")
        self.settings_btn.setText("设置与详情 ▾")
        self.settings_btn.setToolTip("查看预测任务详情、目标层位、显示模式与状态")
        self.settings_btn.setPopupMode(QToolButton.ToolButtonPopupMode.InstantPopup)
        self._build_settings_menu()
        layout.addWidget(self.settings_btn)

        # 4. 状态标签 (Status Indicator)
        status_prefix = QLabel("状态:")
        status_prefix.setObjectName("WorkFieldLabel")
        layout.addWidget(status_prefix)

        self.status_value = QLabel("—")
        self.status_value.setObjectName("WorkFieldValue")
        layout.addWidget(self.status_value)

        # 5. 弹性伸缩
        layout.addStretch(1)

        # 6. 操作按钮 (Action Buttons)
        self.demo_btn = QPushButton("运行演示预测")
        self.demo_btn.setObjectName("SecondaryButton")
        self.demo_btn.setToolTip(
            "显式演示模式：运行 DemoModelProvider（合成数据，非科学预测）"
        )
        self.demo_btn.clicked.connect(self.demo_requested.emit)
        layout.addWidget(self.demo_btn)

        self.run_btn = QPushButton("运行预测")
        self.run_btn.setObjectName("PrimaryButton")
        self.run_btn.setToolTip(
            "通过 ModelRegistry 解析生产模型后运行科学预测；"
            "未配置生产模型时不会自动运行 mock"
        )
        self.run_btn.clicked.connect(self.run_requested.emit)
        layout.addWidget(self.run_btn)

    def _build_settings_menu(self) -> None:
        """Create the popup menu containing context details and mode switches."""
        self.settings_menu = QMenu(self)
        self.settings_menu.setObjectName("SeismicSettingsMenu")

        # Context details card widget embedded inside QWidgetAction
        details_widget = QFrame()
        details_widget.setObjectName("SeismicSettingsDetailsCard")
        grid = QGridLayout(details_widget)
        grid.setContentsMargins(tokens.SPACE_3, tokens.SPACE_2, tokens.SPACE_3, tokens.SPACE_2)
        grid.setSpacing(tokens.SPACE_2)

        # Task
        task_lbl = QLabel("任务:")
        task_lbl.setObjectName("WorkFieldLabel")
        self.task_value = QLabel("未选择预测任务")
        self.task_value.setObjectName("WorkFieldValue")
        grid.addWidget(task_lbl, 0, 0)
        grid.addWidget(self.task_value, 0, 1)

        # Horizon
        horizon_lbl = QLabel("层位:")
        horizon_lbl.setObjectName("WorkFieldLabel")
        self.horizon_value = QLabel("—")
        self.horizon_value.setObjectName("WorkFieldValue")
        grid.addWidget(horizon_lbl, 1, 0)
        grid.addWidget(self.horizon_value, 1, 1)

        # Attribute
        attr_lbl = QLabel("当前属性:")
        attr_lbl.setObjectName("WorkFieldLabel")
        self.attribute_value = QLabel("振幅")
        self.attribute_value.setObjectName("WorkFieldValue")
        grid.addWidget(attr_lbl, 2, 0)
        grid.addWidget(self.attribute_value, 2, 1)

        # Display Mode
        mode_lbl = QLabel("显示模式:")
        mode_lbl.setObjectName("WorkFieldLabel")
        self.mode_value = QLabel("vd")
        self.mode_value.setObjectName("WorkFieldValue")
        grid.addWidget(mode_lbl, 3, 0)
        grid.addWidget(self.mode_value, 3, 1)

        # Volume Shape
        shape_lbl = QLabel("体数据维度:")
        shape_lbl.setObjectName("WorkFieldLabel")
        self.shape_value = QLabel("—")
        self.shape_value.setObjectName("WorkFieldValue")
        grid.addWidget(shape_lbl, 4, 0)
        grid.addWidget(self.shape_value, 4, 1)

        # Output Nature
        mock_lbl = QLabel("输出性质:")
        mock_lbl.setObjectName("WorkFieldLabel")
        self.mock_value = QLabel("—")
        self.mock_value.setObjectName("WorkFieldValue")
        grid.addWidget(mock_lbl, 5, 0)
        grid.addWidget(self.mock_value, 5, 1)

        card_action = QWidgetAction(self.settings_menu)
        card_action.setDefaultWidget(details_widget)
        self.settings_menu.addAction(card_action)

        self.settings_menu.addSeparator()

        # Display Mode Submenu
        mode_menu = self.settings_menu.addMenu("切换显示模式")
        self._mode_group = QActionGroup(self)
        for mode_name in ("vd", "wiggle"):
            act = QAction(mode_name, self)
            act.setCheckable(True)
            if mode_name == "vd":
                act.setChecked(True)
            act.triggered.connect(lambda _c=False, m=mode_name: self._on_mode_action_triggered(m))
            self._mode_group.addAction(act)
            mode_menu.addAction(act)

        # Attribute Categories Submenu
        attr_menu = self.settings_menu.addMenu("切换地震属性")
        for group_label, labels in _ATTRIBUTE_GROUPS:
            group_sub = attr_menu.addMenu(group_label)
            for label in labels:
                act = QAction(label, self)
                act.triggered.connect(lambda _c=False, l=label: self._on_attr_action_triggered(l))
                group_sub.addAction(act)

        self.settings_btn.setMenu(self.settings_menu)

    def _on_attribute_combo_changed(self, text: str) -> None:
        if not self._suppress_signals and text:
            self.attribute_value.setText(text)
            self.attribute_changed.emit(text)

    def _on_mode_action_triggered(self, mode: str) -> None:
        self.mode_value.setText(mode)
        self.display_mode_changed.emit(mode)

    def _on_attr_action_triggered(self, label: str) -> None:
        self.set_selected_attribute(label)
        self.attribute_changed.emit(label)

    def set_selected_attribute(self, label: str) -> None:
        """Synchronize the attribute combo and label without recursion."""
        text = str(label or "").strip()
        if not text:
            return
        self.attribute_value.setText(text)
        self._suppress_signals = True
        idx = self.attribute_combo.findText(text)
        if idx >= 0:
            self.attribute_combo.setCurrentIndex(idx)
        self._suppress_signals = False

    def set_context(
        self,
        task,
        horizon: str,
        attribute: str,
        display_mode: str,
        volume_shape: tuple[int, int, int] | None = None,
        mock_nature: str | None = None,
    ) -> None:
        self.task_value.setText(field_value(task, "name", "") or "未选择预测任务")
        self.horizon_value.setText(str(horizon or "—"))
        self.set_selected_attribute(attribute or "振幅")
        mode = str(display_mode or "vd")
        self.mode_value.setText(mode)
        for act in self._mode_group.actions():
            if act.text() == mode:
                act.setChecked(True)
        if volume_shape is not None:
            self.shape_value.setText(" × ".join(str(v) for v in volume_shape) if volume_shape else "—")
        if mock_nature is not None:
            self.mock_value.setText(str(mock_nature))

    def set_inferring(self, busy: bool) -> None:
        """Enable/disable the run/demo actions for the duration of a run."""
        self._inferring = bool(busy)
        self.run_btn.setEnabled(not self._inferring)
        self.demo_btn.setEnabled(not self._inferring)
        if busy:
            self.status_value.setText("推断中…")

    def set_status(self, text: str) -> None:
        """Show an async run outcome in-page (no modal dialogs, #897)."""
        self.status_value.setText(str(text or "—"))
